Objects/unicodeobject.c

```diff
@@ -2272,34 +2269,74 @@ PyUnicode_FromString(const char *u)
     return PyUnicode_DecodeUTF8Stateful(u, (Py_ssize_t)size, NULL, NULL);
 }
 
+
 PyObject *
 _PyUnicode_FromId(_Py_Identifier *id)
 {
-    if (!id->object) {
-        id->object = PyUnicode_DecodeUTF8Stateful(id->string,
-                                                  strlen(id->string),
-                                                  NULL, NULL);
-        if (!id->object)
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    struct _Py_unicode_ids *ids = &interp->unicode.ids;
+
+    if (id->index < 0) {
```

Comment:
"It is easier to initialize static variables with 0"

You cannot initialize a _Py_Identifier to zeros: the string field must be set. IMO it's reasonable to require users of this API to use _Py_static_string_init(), _Py_static_string() or _Py_IDENTIFIER() macro and don't initialize members manually.

Reply:
I am not sure, but it looks to me that it may be cheaper for the loader to initialize non-constant static variable with zeros than with some other value. For non-zero value it needs to copy it from some place (read-only) to the read-write data area. For zero value it can just keep initial zeros (if the data page is filled by zeros on allocation). It can save few bytes in executable file and address space and few tacts at program start time.

Reply:
> I am not sure, but it looks to me that it may be cheaper for the loader to initialize non-constant static variable with zeros than with some other value.

_Py_Identifier is made of 2 members: string (non-NULL constant value) and index (set once at runtime, get many times). The structure is never initialized to zeros, even if we change the default "not set" index to 0.